src/tls_gnutls.c

```diff
@@ -58,6 +120,15 @@ tls_t *tls_new(xmpp_conn_t *conn)
         gnutls_certificate_allocate_credentials(&tls->cred);
         tls_set_credentials(tls, CAFILE);
 
+        if (conn->tls_client_cert && conn->tls_client_key) {
+            gnutls_certificate_set_x509_key_file(
+                tls->cred, conn->tls_client_cert, conn->tls_client_key,
+                GNUTLS_X509_FMT_PEM);
+            if (conn->tls_client_auth_mandatory)
+                gnutls_certificate_server_set_request(tls->session,
+                                                      GNUTLS_CERT_REQUIRE);
```

Comment:
I doubt that this logic would work, because `gnutls_certificate_server_set_request` should only affect the server's behavior (and in TLS, it's always the server that initiates client authentication). To check if the client-auth has happend, I guess you could use either a handshake callback or `gnutls_certificate_client_get_request_status`, after the handshake is complete.

Reply:
thanks! that's exactly the feedback I hoped for :)

@pasis I'll adapt this behavior (most likely on Monday)